src/test/benchmarks/io/vertx/benchmarks/SynchronizationBenchmark.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.benchmarks;
+
+import io.vertx.core.net.impl.pool.LockSynchronization;
+import io.vertx.core.net.impl.pool.NonBlockingSynchronization1;
+import io.vertx.core.net.impl.pool.NonBlockingSynchronization2;
+import io.vertx.core.net.impl.pool.Synchronization;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * @author Thomas Segismont
+ * @author slinkydeveloper
+ */
+@State(Scope.Benchmark)
+@Warmup(iterations = 20, time = 200, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 200, timeUnit = MILLISECONDS)
+@Threads(2)
+public class SynchronizationBenchmark extends BenchmarkBase {
+
+  private Synchronization<Object> synchronization1;
+  private Synchronization<Object> synchronization2;
+  private Synchronization.Action<Object> action;
+
+  private CountDownLatch latch = new CountDownLatch(1);
+
+  @Setup
+  public void setup() throws Exception {
+    synchronization1 = new NonBlockingSynchronization1<>(new Object());
+    synchronization2 = new NonBlockingSynchronization2<>(new Object());
+    action = state -> {
+      Blackhole.consumeCPU(0);
+      return null;
+    };
+    CountDownLatch l = new CountDownLatch(2);
+    new Thread(() -> {
+      synchronization1.execute(state -> {
+        l.countDown();
+        try {
+          latch.await();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+        return null;
+      });
+    }).start();
+    new Thread(() -> {
+      synchronization2.execute(state -> {
+        l.countDown();
+        try {
+          latch.await();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+        return null;
+      });
+    }).start();
+    l.await(20, TimeUnit.SECONDS);
+  }
+
+  @TearDown
+  public void tearDown() {
+    latch.countDown();
+  }
+
+  @Benchmark
+  public void lock1() {
+    synchronization1.execute(action);
```

Comment:
We're interested when a specific action is going to be completed too, because, if not, we just risk to measure the cost of submission and just using an unbunded q to accumulate tasks would perform best then any possible implementation (until OOM).
In addition, this bench need repeatable condition and using RTT can achieve steady state.
I would change it like this:
```java
@Param
int burstSize;

@Param
int producerDelay;

// consumerDelay is used on action execution to simulare real CPU work
@Param 
int consumerDelay;

// create an enum that use this to inject a specific synchronization impl instead of using N different methods
@Param
private String lockType;

// PerThreadState is a Thread scoped state that hold resettable actions
@Benchmark
public void lock(PerThreadState state) {
   if (producerDelay > 0) {
      BlackHole.consumeCPU(producerDelay);
   }
   // this is going to execute all actions and spin wait completion -> can use a single AtomicLong somewhere
   state.executeActionsAndWaitCompletion(sync);
}
```

Reply:
it is actually the intention to measure the submission cost to compare the two implementations. I think what you describe can be another benchmark.

Reply:
Yep but the submission cost depends a lot if the queue is streaming or is near empty...and that's not reproducible because depends by the backlog of tasks...
Benchs like this work better while having a steady state and the RTT burst ones can help to reach it, that's why I have proposed it :)